<a href="https://colab.research.google.com/github/ABHIRAM199/my_practice/blob/main/Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Agenda

### 1. Setting Up the Environment in Google Colab
   - Downloading and Organizing CSV Files
   - Importing Pandas in Google Colab
   - Reading and Understanding Data


### 2. Pandas Tutorial
  - Data Frame(DF)
     - Data Frame Operations
  - Series
     - Series Operations
  - Understanding loc and iloc
  - Basic Operations in Pandas
      - Filtering Data
      - Grouping Data
      - Sorting Data
      - String Operations
      - Resetting Index
  - Restructuring Data with Melt
  - Applying Custom Functions in Pandas

### 3. Creating Series and DataFrames from Scratch
   - Creating a Series
   - Creating a DataFrame
   - Concatenating a Dataframe
   

### 4. Dealing with Missing Values
   - General Knowledge on Reading CSV files
   - Understanding Missing Values
   - Strategies for Handling Missing Data
      - Deletion Methods
      - Imputation Methods
      
### 5.Exploratory Data Analysis (EDA)
   - Data Cleaning and Manipulation
   - Working with DateTime   
   


# Setting Up Your Environment




## Downloading and Organizing CSV Files
-  Create a folder in Google Collab with the name **DataScienceCourse**, making it easy to manage and access notebooks.
- Import the given **[gapminder.csv](https://www.kaggle.com/datasets/albertovidalrod/gapminder-dataset)**, **[disease.csv](https://github.com/kb22/Heart-Disease-Prediction/raw/master/dataset.csv)** and **[weather.csv](https://github.com/yesdhruv/csv_files/blob/master/weather.csv)** files and upload it to  Google Collab in the DataScienceCourse Folder as we will be using these.

## Importing Pandas in Google Colab
Make sure that Pandas is available in your Colab environment. If not we can use the below command:




In [ ]:
!pip install pandas

After installing let's import pandas. It's a standard practice to use **pd** for using the pandas functions.

In [ ]:
import pandas as pd

## Reading and Understanding Data
* Overview of the Gapminder Dataset
 * The Gapminder dataset is a comprehensive repository of global development metrics, offering a nuanced view of societal progress. Renowned for its use in captivating TED Talks by Hans Rosling, this dataset encapsulates key indicators such as life expectancy, GDP per capita, and population across diverse countries and years.

- Import the 'gapminder.csv' file directly from github url or you do the upload method taht weit using the steps discussed above





In [ ]:
# PANDAS:

# excel
# text
# csv
# json
# parquet
# ...
# ...
# ...

# READING FROM OUR LOCAL MACHINE....

# pandas_dataframe=pd.read_excel("path of the file.xlsx")
# pandas_dataframe=pd.read_csv(".csv")
# pandas_dataframe=pd.read_json(".json")


# WE COULD ALSO READ DATA FROM PUBLIC LOCATIONS--> raw_github

In [ ]:
# Use the correct raw URL for the CSV file on GitHub
csv_url = 'https://raw.githubusercontent.com/yesdhruv/gapminder.csv/master/gapminder_data_graphs.csv'


In [ ]:
df = pd.read_csv(csv_url)
# Display the DataFrame
df

country continent  year  life_exp  hdi_index  co2_consump     gdp  \
0     Afghanistan      Asia  1998      53.3      0.344       0.0522     NaN   
1     Afghanistan      Asia  1999      54.7      0.348       0.0402     NaN   
2     Afghanistan      Asia  2000      54.7      0.350       0.0370     NaN   
3     Afghanistan      Asia  2001      54.8      0.353       0.0376     NaN   
4     Afghanistan      Asia  2002      55.5      0.384       0.0471   333.0   
...           ...       ...   ...       ...        ...          ...     ...   
3670     Zimbabwe    Africa  2014      58.0      0.547       0.8810  1440.0   
3671     Zimbabwe    Africa  2015      58.6      0.553       0.8810  1450.0   
3672     Zimbabwe    Africa  2016      59.2      0.558       0.7710  1430.0   
3673     Zimbabwe    Africa  2017      59.9      0.563       0.8450  1480.0   
3674     Zimbabwe    Africa  2018      60.6      0.569       0.8500  1510.0   

      services  
0         24.4  
1         24.6  
2         24.7  
3         24.7  
4         25.6  
...        ...  
3670      25.4  
3671      25.7  
3672      26.1  
3673      26.6  
3674      27.2  

[3675 rows x 8 columns]

In [ ]:
df.shape

(3675, 8)

The above code will display a 2-dimensional labeled structure with diverse data types and as we can see the 3675 rows with 8 columns which represents the size of the data frame.
Pandas give each row a number (starting from 0) in the 'gapminder.csv' DataFrame.
The table shows column names at the top, including data like country names and numbers.

# Pandas Tutorial
Pandas is a powerful and user-friendly Python library that simplifies data manipulation and analysis. Pandas helps in organizing, cleaning, and analyzing the data seamlessly.

In [ ]:
df.loc[2,'life_exp']

54.7

In [ ]:
df.iloc[2,2]

2000

In [ ]:
df

country continent  year  life_exp  hdi_index  co2_consump     gdp  \
0     Afghanistan      Asia  1998      53.3      0.344       0.0522     NaN   
1     Afghanistan      Asia  1999      54.7      0.348       0.0402     NaN   
2     Afghanistan      Asia  2000      54.7      0.350       0.0370     NaN   
3     Afghanistan      Asia  2001      54.8      0.353       0.0376     NaN   
4     Afghanistan      Asia  2002      55.5      0.384       0.0471   333.0   
...           ...       ...   ...       ...        ...          ...     ...   
3670     Zimbabwe    Africa  2014      58.0      0.547       0.8810  1440.0   
3671     Zimbabwe    Africa  2015      58.6      0.553       0.8810  1450.0   
3672     Zimbabwe    Africa  2016      59.2      0.558       0.7710  1430.0   
3673     Zimbabwe    Africa  2017      59.9      0.563       0.8450  1480.0   
3674     Zimbabwe    Africa  2018      60.6      0.569       0.8500  1510.0   

      services  
0         24.4  
1         24.6  
2         24.7  
3         24.7  
4         25.6  
...        ...  
3670      25.4  
3671      25.7  
3672      26.1  
3673      26.6  
3674      27.2  

[3675 rows x 8 columns]

 ### Data Frame (DF)
  In pandas, a Data Frame represents a table and is the default storage structure for tabular data. Tables created by reading CSV files, like DF = pd.read_csv('file.csv'), are stored as Data Frames.

* Usage

   * Accessing Rows: You can access rows using df.loc[]. For example, df.loc[0] retrieves the first row of the DataFrame df.
   * Row as Series: When accessing a single row, it is returned as a Series, not a DataFrame

**Note:** Data Frames can handle diverse data types, providing versatility beyond numeric values


#### DataFrame Operations

1. Creating a DataFrame- Data is read from a CSV file and stored in a variable, typically named DF (short for DataFrame).



In [ ]:
# storing the file content in DF variable
import pandas as pd
DF = pd.read_csv(csv_url)

2. Printing the DataFrame:
Writing DF in a cell and executing it will display the DataFrame in Google Colab.

In [ ]:
DF

country continent  year  life_exp  hdi_index  co2_consump     gdp  \
0     Afghanistan      Asia  1998      53.3      0.344       0.0522     NaN   
1     Afghanistan      Asia  1999      54.7      0.348       0.0402     NaN   
2     Afghanistan      Asia  2000      54.7      0.350       0.0370     NaN   
3     Afghanistan      Asia  2001      54.8      0.353       0.0376     NaN   
4     Afghanistan      Asia  2002      55.5      0.384       0.0471   333.0   
...           ...       ...   ...       ...        ...          ...     ...   
3670     Zimbabwe    Africa  2014      58.0      0.547       0.8810  1440.0   
3671     Zimbabwe    Africa  2015      58.6      0.553       0.8810  1450.0   
3672     Zimbabwe    Africa  2016      59.2      0.558       0.7710  1430.0   
3673     Zimbabwe    Africa  2017      59.9      0.563       0.8450  1480.0   
3674     Zimbabwe    Africa  2018      60.6      0.569       0.8500  1510.0   

      services  
0         24.4  
1         24.6  
2         24.7  
3         24.7  
4         25.6  
...        ...  
3670      25.4  
3671      25.7  
3672      26.1  
3673      26.6  
3674      27.2  

[3675 rows x 8 columns]

3. Viewing Top Rows (head):
DF.head() displays the top five rows of the DataFrame by default.
Inorder to display specific number of rows we can pass it to the header parameter.


In [ ]:
# UNDERSTANDING THE DATA

In [ ]:
DF.shape

(3675, 8)

In [ ]:
DF.head(25)

country continent  year  life_exp  hdi_index  co2_consump     gdp  \
0   Afghanistan      Asia  1998      53.3      0.344       0.0522     NaN   
1   Afghanistan      Asia  1999      54.7      0.348       0.0402     NaN   
2   Afghanistan      Asia  2000      54.7      0.350       0.0370     NaN   
3   Afghanistan      Asia  2001      54.8      0.353       0.0376     NaN   
4   Afghanistan      Asia  2002      55.5      0.384       0.0471   333.0   
5   Afghanistan      Asia  2003      56.5      0.393       0.0509   346.0   
6   Afghanistan      Asia  2004      57.1      0.409       0.0368   336.0   
7   Afghanistan      Asia  2005      57.6      0.418       0.0515   360.0   
8   Afghanistan      Asia  2006      58.0      0.429       0.0622   368.0   
9   Afghanistan      Asia  2007      58.5      0.447       0.0838   409.0   
10  Afghanistan      Asia  2008      59.2      0.447       0.1520   415.0   
11  Afghanistan      Asia  2009      59.9      0.460       0.2380   492.0   
12  Afghanistan      Asia  2010      60.5      0.472       0.2900   547.0   
13  Afghanistan      Asia  2011      61.0      0.477       0.4060   533.0   
14  Afghanistan      Asia  2012      61.4      0.489       0.3450   580.0   
15  Afghanistan      Asia  2013      61.9      0.496       0.2800   592.0   
16  Afghanistan      Asia  2014      61.9      0.500       0.2530   588.0   
17  Afghanistan      Asia  2015      61.9      0.500       0.2620   578.0   
18  Afghanistan      Asia  2016      62.0      0.502       0.2450   575.0   
19  Afghanistan      Asia  2017      62.9      0.506       0.2470   576.0   
20  Afghanistan      Asia  2018      62.7      0.509       0.2540   569.0   
21      Albania    Europe  1998      74.8      0.655       0.5630  1600.0   
22      Albania    Europe  1999      75.1      0.665       0.9560  1820.0   
23      Albania    Europe  2000      75.4      0.671       0.9660  1960.0   
24      Albania    Europe  2001      76.0      0.678       1.0300  2140.0   

    services  
0       24.4  
1       24.6  
2       24.7  
3       24.7  
4       25.6  
5       25.9  
6       26.1  
7       26.5  
8       26.9  
9       27.7  
10      28.2  
11      29.5  
12      30.7  
13      31.7  
14      32.8  
15      33.8  
16      34.8  
17      35.8  
18      36.9  
19      38.0  
20      38.4  
21      33.0  
22      33.8  
23      34.4  
24      35.0

In [ ]:
DF.head()

country continent  year  life_exp  hdi_index  co2_consump    gdp  \
0  Afghanistan      Asia  1998      53.3      0.344       0.0522    NaN   
1  Afghanistan      Asia  1999      54.7      0.348       0.0402    NaN   
2  Afghanistan      Asia  2000      54.7      0.350       0.0370    NaN   
3  Afghanistan      Asia  2001      54.8      0.353       0.0376    NaN   
4  Afghanistan      Asia  2002      55.5      0.384       0.0471  333.0   

   services  
0      24.4  
1      24.6  
2      24.7  
3      24.7  
4      25.6

In [ ]:
DF.tail()

country continent  year  life_exp  hdi_index  co2_consump     gdp  \
3670  Zimbabwe    Africa  2014      58.0      0.547        0.881  1440.0   
3671  Zimbabwe    Africa  2015      58.6      0.553        0.881  1450.0   
3672  Zimbabwe    Africa  2016      59.2      0.558        0.771  1430.0   
3673  Zimbabwe    Africa  2017      59.9      0.563        0.845  1480.0   
3674  Zimbabwe    Africa  2018      60.6      0.569        0.850  1510.0   

      services  
3670      25.4  
3671      25.7  
3672      26.1  
3673      26.6  
3674      27.2

In [ ]:
DF.tail(25)

country continent  year  life_exp  hdi_index  co2_consump     gdp  \
3650    Zambia    Africa  2015      60.8      0.569        0.271  1340.0   
3651    Zambia    Africa  2016      61.4      0.571        0.296  1350.0   
3652    Zambia    Africa  2017      61.9      0.578        0.304  1350.0   
3653    Zambia    Africa  2018      62.5      0.582        0.302  1370.0   
3654  Zimbabwe    Africa  1998      49.7      0.442        1.200  1780.0   
3655  Zimbabwe    Africa  1999      48.5      0.437        1.330  1760.0   
3656  Zimbabwe    Africa  2000      47.5      0.430        1.160  1700.0   
3657  Zimbabwe    Africa  2001      47.0      0.431        1.050  1720.0   
3658  Zimbabwe    Africa  2002      46.6      0.423        0.996  1560.0   
3659  Zimbabwe    Africa  2003      46.4      0.414        0.886  1290.0   
3660  Zimbabwe    Africa  2004      46.4      0.411        0.785  1210.0   
3661  Zimbabwe    Africa  2005      46.8      0.411        0.887  1140.0   
3662  Zimbabwe    Africa  2006      47.4      0.414        0.853  1090.0   
3663  Zimbabwe    Africa  2007      48.2      0.421        0.803  1040.0   
3664  Zimbabwe    Africa  2008      48.9      0.422        0.624   850.0   
3665  Zimbabwe    Africa  2009      50.2      0.458        0.441   941.0   
3666  Zimbabwe    Africa  2010      52.3      0.482        0.607  1110.0   
3667  Zimbabwe    Africa  2011      54.4      0.499        0.737  1250.0   
3668  Zimbabwe    Africa  2012      56.0      0.525        0.587  1430.0   
3669  Zimbabwe    Africa  2013      57.2      0.537        0.872  1440.0   
3670  Zimbabwe    Africa  2014      58.0      0.547        0.881  1440.0   
3671  Zimbabwe    Africa  2015      58.6      0.553        0.881  1450.0   
3672  Zimbabwe    Africa  2016      59.2      0.558        0.771  1430.0   
3673  Zimbabwe    Africa  2017      59.9      0.563        0.845  1480.0   
3674  Zimbabwe    Africa  2018      60.6      0.569        0.850  1510.0   

      services  
3650      37.1  
3651      38.0  
3652      38.6  
3653      39.3  
3654      27.9  
3655      28.1  
3656      27.8  
3657      27.7  
3658      27.2  
3659      26.5  
3660      26.2  
3661      25.8  
3662      25.4  
3663      25.2  
3664      24.4  
3665      24.8  
3666      25.2  
3667      25.0  
3668      25.4  
3669      25.3  
3670      25.4  
3671      25.7  
3672      26.1  
3673      26.6  
3674      27.2

In [ ]:
# fetching the first top ten rows
DF.head(10)

country continent  year  life_exp  hdi_index  co2_consump    gdp  \
0  Afghanistan      Asia  1998      53.3      0.344       0.0522    NaN   
1  Afghanistan      Asia  1999      54.7      0.348       0.0402    NaN   
2  Afghanistan      Asia  2000      54.7      0.350       0.0370    NaN   
3  Afghanistan      Asia  2001      54.8      0.353       0.0376    NaN   
4  Afghanistan      Asia  2002      55.5      0.384       0.0471  333.0   
5  Afghanistan      Asia  2003      56.5      0.393       0.0509  346.0   
6  Afghanistan      Asia  2004      57.1      0.409       0.0368  336.0   
7  Afghanistan      Asia  2005      57.6      0.418       0.0515  360.0   
8  Afghanistan      Asia  2006      58.0      0.429       0.0622  368.0   
9  Afghanistan      Asia  2007      58.5      0.447       0.0838  409.0   

   services  
0      24.4  
1      24.6  
2      24.7  
3      24.7  
4      25.6  
5      25.9  
6      26.1  
7      26.5  
8      26.9  
9      27.7

4. Viewing Bottom Rows (tail):
DF.tail() shows the last few rows of the DataFrame.
Similar to head, we can specify the number of rows, e.g., DF.tail(3).


In [ ]:
# fetching the last ten rows
DF.tail(10)

country continent  year  life_exp  hdi_index  co2_consump     gdp  \
3665  Zimbabwe    Africa  2009      50.2      0.458        0.441   941.0   
3666  Zimbabwe    Africa  2010      52.3      0.482        0.607  1110.0   
3667  Zimbabwe    Africa  2011      54.4      0.499        0.737  1250.0   
3668  Zimbabwe    Africa  2012      56.0      0.525        0.587  1430.0   
3669  Zimbabwe    Africa  2013      57.2      0.537        0.872  1440.0   
3670  Zimbabwe    Africa  2014      58.0      0.547        0.881  1440.0   
3671  Zimbabwe    Africa  2015      58.6      0.553        0.881  1450.0   
3672  Zimbabwe    Africa  2016      59.2      0.558        0.771  1430.0   
3673  Zimbabwe    Africa  2017      59.9      0.563        0.845  1480.0   
3674  Zimbabwe    Africa  2018      60.6      0.569        0.850  1510.0   

      services  
3665      24.8  
3666      25.2  
3667      25.0  
3668      25.4  
3669      25.3  
3670      25.4  
3671      25.7  
3672      26.1  
3673      26.6  
3674      27.2

5. DataFrame Type (type):
Used to check the DataFrame's type, which should be pandas.core.frame.DataFrame.


In [ ]:
type(DF)

pandas.core.frame.DataFrame

6. DataFrame Shape (shape):
DF.shape returns the dimensions of the DataFrame


In [ ]:
DF.shape

(3675, 8)

7. DataFrame Info (info):
DF.info() provides a summary of the DataFrame, including the number of non-null values and data types of each column.


In [ ]:
DF.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3675 entries, 0 to 3674
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   country      3675 non-null   object 
 1   continent    3675 non-null   object 
 2   year         3675 non-null   int64  
 3   life_exp     3675 non-null   float64
 4   hdi_index    3563 non-null   float64
 5   co2_consump  3671 non-null   float64
 6   gdp          3633 non-null   float64
 7   services     3675 non-null   float64
dtypes: float64(5), int64(1), object(2)
memory usage: 229.8+ KB


**Note:** Use .shape and .info() methods to understand the Dataframe's structure and data types.


8. DataFrame Describe (describe):
DF.describe() offers a statistical summary of numerical columns, including count, mean, standard deviation, min/max values, and percentiles.


In [ ]:
DF.describe().year[3]

1998.0

In [ ]:
DF.describe()

year     life_exp    hdi_index  co2_consump            gdp  \
count  3675.000000  3675.000000  3563.000000  3671.000000    3633.000000   
mean   2008.000000    69.849306     0.674864     4.712731   11966.053675   
std       6.056125     8.886563     0.164834     6.567435   17105.787953   
min    1998.000000    32.500000     0.255000     0.015900     238.000000   
25%    2003.000000    63.900000     0.537500     0.560500    1470.000000   
50%    2008.000000    71.700000     0.699000     2.250000    4280.000000   
75%    2013.000000    76.400000     0.805000     6.615000   13600.000000   
max    2018.000000    84.800000     0.956000    67.100000  105000.000000   

          services  
count  3675.000000  
mean     51.248705  
std      18.312501  
min       5.590000  
25%      37.600000  
50%      52.900000  
75%      65.700000  
max      88.500000

9. Accessing DataFrame Columns:
Use DF['column_name'] to access a specific column.
For multiple columns, use a list inside the brackets, e.g., DF[['column1', 'column2']].


In [ ]:
DF

country continent  year  life_exp  hdi_index  co2_consump     gdp  \
0     Afghanistan      Asia  1998      53.3      0.344       0.0522     NaN   
1     Afghanistan      Asia  1999      54.7      0.348       0.0402     NaN   
2     Afghanistan      Asia  2000      54.7      0.350       0.0370     NaN   
3     Afghanistan      Asia  2001      54.8      0.353       0.0376     NaN   
4     Afghanistan      Asia  2002      55.5      0.384       0.0471   333.0   
...           ...       ...   ...       ...        ...          ...     ...   
3670     Zimbabwe    Africa  2014      58.0      0.547       0.8810  1440.0   
3671     Zimbabwe    Africa  2015      58.6      0.553       0.8810  1450.0   
3672     Zimbabwe    Africa  2016      59.2      0.558       0.7710  1430.0   
3673     Zimbabwe    Africa  2017      59.9      0.563       0.8450  1480.0   
3674     Zimbabwe    Africa  2018      60.6      0.569       0.8500  1510.0   

      services  
0         24.4  
1         24.6  
2         24.7  
3         24.7  
4         25.6  
...        ...  
3670      25.4  
3671      25.7  
3672      26.1  
3673      26.6  
3674      27.2  

[3675 rows x 8 columns]

In [ ]:
DF['hdi_index']

0       0.344
1       0.348
2       0.350
3       0.353
4       0.384
        ...  
3670    0.547
3671    0.553
3672    0.558
3673    0.563
3674    0.569
Name: hdi_index, Length: 3675, dtype: float64

In [ ]:
DF[['continent','life_exp','hdi_index']]

continent  life_exp  hdi_index
0         Asia      53.3      0.344
1         Asia      54.7      0.348
2         Asia      54.7      0.350
3         Asia      54.8      0.353
4         Asia      55.5      0.384
...        ...       ...        ...
3670    Africa      58.0      0.547
3671    Africa      58.6      0.553
3672    Africa      59.2      0.558
3673    Africa      59.9      0.563
3674    Africa      60.6      0.569

[3675 rows x 3 columns]

In [ ]:
DF['continent']

0         Asia
1         Asia
2         Asia
3         Asia
4         Asia
         ...  
3670    Africa
3671    Africa
3672    Africa
3673    Africa
3674    Africa
Name: continent, Length: 3675, dtype: object

In [ ]:
#accesing single column
DF['gdp']

0          NaN
1          NaN
2          NaN
3          NaN
4        333.0
         ...  
3670    1440.0
3671    1450.0
3672    1430.0
3673    1480.0
3674    1510.0
Name: gdp, Length: 3675, dtype: float64

In [ ]:
#accesing multiples column
DF[['gdp','services']]

gdp  services
0        NaN      24.4
1        NaN      24.6
2        NaN      24.7
3        NaN      24.7
4      333.0      25.6
...      ...       ...
3670  1440.0      25.4
3671  1450.0      25.7
3672  1430.0      26.1
3673  1480.0      26.6
3674  1510.0      27.2

[3675 rows x 2 columns]

**Note:** we can also print the list of columns headers(title) by using the below command. It is beacuse column is an attribute.

In [ ]:
DF.columns

Index(['country', 'continent', 'year', 'life_exp', 'hdi_index', 'co2_consump',
       'gdp', 'services'],
      dtype='object')

In [ ]:
# DATAFRAMES can be like your DICTIONARIES, because they also consist of your KEYS and VALUES.

Dataframes can be used as keys using **df.keys()**. It serves the same purpose like dictinary data *structure*.

In [ ]:
df.values

array([['Afghanistan', 'Asia', 1998, ..., 0.0522, nan, 24.4],
       ['Afghanistan', 'Asia', 1999, ..., 0.0402, nan, 24.6],
       ['Afghanistan', 'Asia', 2000, ..., 0.037, nan, 24.7],
       ...,
       ['Zimbabwe', 'Africa', 2016, ..., 0.771, 1430.0, 26.1],
       ['Zimbabwe', 'Africa', 2017, ..., 0.845, 1480.0, 26.6],
       ['Zimbabwe', 'Africa', 2018, ..., 0.85, 1510.0, 27.2]],
      dtype=object)

In [ ]:
df.keys()

Index(['country', 'continent', 'year', 'life_exp', 'hdi_index', 'co2_consump',
       'gdp', 'services'],
      dtype='object')


Additionally, we can perform other sets of *operations* by combining the different methods

In [ ]:
#print the first five values form column named country
DF['country'].head()

0    Afghanistan
1    Afghanistan
2    Afghanistan
3    Afghanistan
4    Afghanistan
Name: country, dtype: object

In [ ]:
#print the first five values form columns named country and life_exp
DF[['country','life_exp']].head()

country  life_exp
0  Afghanistan      53.3
1  Afghanistan      54.7
2  Afghanistan      54.7
3  Afghanistan      54.8
4  Afghanistan      55.5

10. Dropping a Column (drop):


*   To remove a column, use DF.drop('column_name', axis=1).

In [ ]:
DF

country continent  year  life_exp  hdi_index  co2_consump     gdp  \
0     Afghanistan      Asia  1998      53.3      0.344       0.0522     NaN   
1     Afghanistan      Asia  1999      54.7      0.348       0.0402     NaN   
2     Afghanistan      Asia  2000      54.7      0.350       0.0370     NaN   
3     Afghanistan      Asia  2001      54.8      0.353       0.0376     NaN   
4     Afghanistan      Asia  2002      55.5      0.384       0.0471   333.0   
...           ...       ...   ...       ...        ...          ...     ...   
3670     Zimbabwe    Africa  2014      58.0      0.547       0.8810  1440.0   
3671     Zimbabwe    Africa  2015      58.6      0.553       0.8810  1450.0   
3672     Zimbabwe    Africa  2016      59.2      0.558       0.7710  1430.0   
3673     Zimbabwe    Africa  2017      59.9      0.563       0.8450  1480.0   
3674     Zimbabwe    Africa  2018      60.6      0.569       0.8500  1510.0   

      services  
0         24.4  
1         24.6  
2         24.7  
3         24.7  
4         25.6  
...        ...  
3670      25.4  
3671      25.7  
3672      26.1  
3673      26.6  
3674      27.2  

[3675 rows x 8 columns]

In [ ]:
DF.columns

Index(['country', 'continent', 'year', 'life_exp', 'hdi_index', 'co2_consump',
       'gdp', 'services'],
      dtype='object')

In [ ]:
# DF.drop('continent')

In [ ]:
DF.drop('continent',axis=1)

country  year  life_exp  hdi_index  co2_consump     gdp  services
0     Afghanistan  1998      53.3      0.344       0.0522     NaN      24.4
1     Afghanistan  1999      54.7      0.348       0.0402     NaN      24.6
2     Afghanistan  2000      54.7      0.350       0.0370     NaN      24.7
3     Afghanistan  2001      54.8      0.353       0.0376     NaN      24.7
4     Afghanistan  2002      55.5      0.384       0.0471   333.0      25.6
...           ...   ...       ...        ...          ...     ...       ...
3670     Zimbabwe  2014      58.0      0.547       0.8810  1440.0      25.4
3671     Zimbabwe  2015      58.6      0.553       0.8810  1450.0      25.7
3672     Zimbabwe  2016      59.2      0.558       0.7710  1430.0      26.1
3673     Zimbabwe  2017      59.9      0.563       0.8450  1480.0      26.6
3674     Zimbabwe  2018      60.6      0.569       0.8500  1510.0      27.2

[3675 rows x 7 columns]

In [ ]:
DF

country continent  year  life_exp  hdi_index  co2_consump     gdp  \
0     Afghanistan      Asia  1998      53.3      0.344       0.0522     NaN   
1     Afghanistan      Asia  1999      54.7      0.348       0.0402     NaN   
2     Afghanistan      Asia  2000      54.7      0.350       0.0370     NaN   
3     Afghanistan      Asia  2001      54.8      0.353       0.0376     NaN   
4     Afghanistan      Asia  2002      55.5      0.384       0.0471   333.0   
...           ...       ...   ...       ...        ...          ...     ...   
3670     Zimbabwe    Africa  2014      58.0      0.547       0.8810  1440.0   
3671     Zimbabwe    Africa  2015      58.6      0.553       0.8810  1450.0   
3672     Zimbabwe    Africa  2016      59.2      0.558       0.7710  1430.0   
3673     Zimbabwe    Africa  2017      59.9      0.563       0.8450  1480.0   
3674     Zimbabwe    Africa  2018      60.6      0.569       0.8500  1510.0   

      services  
0         24.4  
1         24.6  
2         24.7  
3         24.7  
4         25.6  
...        ...  
3670      25.4  
3671      25.7  
3672      26.1  
3673      26.6  
3674      27.2  

[3675 rows x 8 columns]

In [ ]:
 DF.drop('year', axis=1)

country continent  life_exp  hdi_index  co2_consump     gdp  \
0     Afghanistan      Asia      53.3      0.344       0.0522     NaN   
1     Afghanistan      Asia      54.7      0.348       0.0402     NaN   
2     Afghanistan      Asia      54.7      0.350       0.0370     NaN   
3     Afghanistan      Asia      54.8      0.353       0.0376     NaN   
4     Afghanistan      Asia      55.5      0.384       0.0471   333.0   
...           ...       ...       ...        ...          ...     ...   
3670     Zimbabwe    Africa      58.0      0.547       0.8810  1440.0   
3671     Zimbabwe    Africa      58.6      0.553       0.8810  1450.0   
3672     Zimbabwe    Africa      59.2      0.558       0.7710  1430.0   
3673     Zimbabwe    Africa      59.9      0.563       0.8450  1480.0   
3674     Zimbabwe    Africa      60.6      0.569       0.8500  1510.0   

      services  
0         24.4  
1         24.6  
2         24.7  
3         24.7  
4         25.6  
...        ...  
3670      25.4  
3671      25.7  
3672      26.1  
3673      26.6  
3674      27.2  

[3675 rows x 7 columns]

*   To reflect this change in the original DataFrame, use inplace=True or reassign it like DF = DF.drop('column_name', axis=1).

In [ ]:
df.drop()

ValueError: ignored

In [ ]:
# DF = DF.drop('year', axis=1)

DF

In [ ]:
DF.drop(['life_exp','hdi_index'],axis=1)

country continent  year  co2_consump     gdp  services
0     Afghanistan      Asia  1998       0.0522     NaN      24.4
1     Afghanistan      Asia  1999       0.0402     NaN      24.6
2     Afghanistan      Asia  2000       0.0370     NaN      24.7
3     Afghanistan      Asia  2001       0.0376     NaN      24.7
4     Afghanistan      Asia  2002       0.0471   333.0      25.6
...           ...       ...   ...          ...     ...       ...
3670     Zimbabwe    Africa  2014       0.8810  1440.0      25.4
3671     Zimbabwe    Africa  2015       0.8810  1450.0      25.7
3672     Zimbabwe    Africa  2016       0.7710  1430.0      26.1
3673     Zimbabwe    Africa  2017       0.8450  1480.0      26.6
3674     Zimbabwe    Africa  2018       0.8500  1510.0      27.2

[3675 rows x 6 columns]

11. Resetting the DataFrame:
To revert to the original DataFrame, reread the CSV file.


In [ ]:
DF = pd.read_csv(csv_url)
DF

country continent  year  life_exp  hdi_index  co2_consump     gdp  \
0     Afghanistan      Asia  1998      53.3      0.344       0.0522     NaN   
1     Afghanistan      Asia  1999      54.7      0.348       0.0402     NaN   
2     Afghanistan      Asia  2000      54.7      0.350       0.0370     NaN   
3     Afghanistan      Asia  2001      54.8      0.353       0.0376     NaN   
4     Afghanistan      Asia  2002      55.5      0.384       0.0471   333.0   
...           ...       ...   ...       ...        ...          ...     ...   
3670     Zimbabwe    Africa  2014      58.0      0.547       0.8810  1440.0   
3671     Zimbabwe    Africa  2015      58.6      0.553       0.8810  1450.0   
3672     Zimbabwe    Africa  2016      59.2      0.558       0.7710  1430.0   
3673     Zimbabwe    Africa  2017      59.9      0.563       0.8450  1480.0   
3674     Zimbabwe    Africa  2018      60.6      0.569       0.8500  1510.0   

      services  
0         24.4  
1         24.6  
2         24.7  
3         24.7  
4         25.6  
...        ...  
3670      25.4  
3671      25.7  
3672      26.1  
3673      26.6  
3674      27.2  

[3675 rows x 8 columns]

In [ ]:
df.values

array([['Afghanistan', 'Asia', 1998, ..., 0.0522, nan, 24.4],
       ['Afghanistan', 'Asia', 1999, ..., 0.0402, nan, 24.6],
       ['Afghanistan', 'Asia', 2000, ..., 0.037, nan, 24.7],
       ...,
       ['Zimbabwe', 'Africa', 2016, ..., 0.771, 1430.0, 26.1],
       ['Zimbabwe', 'Africa', 2017, ..., 0.845, 1480.0, 26.6],
       ['Zimbabwe', 'Africa', 2018, ..., 0.85, 1510.0, 27.2]],
      dtype=object)

In [ ]:
#dropna, which drops or removes rows with 1 or more null value.
df.dropna()

country continent  year  life_exp  hdi_index  co2_consump     gdp  \
4     Afghanistan      Asia  2002      55.5      0.384       0.0471   333.0   
5     Afghanistan      Asia  2003      56.5      0.393       0.0509   346.0   
6     Afghanistan      Asia  2004      57.1      0.409       0.0368   336.0   
7     Afghanistan      Asia  2005      57.6      0.418       0.0515   360.0   
8     Afghanistan      Asia  2006      58.0      0.429       0.0622   368.0   
...           ...       ...   ...       ...        ...          ...     ...   
3670     Zimbabwe    Africa  2014      58.0      0.547       0.8810  1440.0   
3671     Zimbabwe    Africa  2015      58.6      0.553       0.8810  1450.0   
3672     Zimbabwe    Africa  2016      59.2      0.558       0.7710  1430.0   
3673     Zimbabwe    Africa  2017      59.9      0.563       0.8450  1480.0   
3674     Zimbabwe    Africa  2018      60.6      0.569       0.8500  1510.0   

      services  
4         25.6  
5         25.9  
6         26.1  
7         26.5  
8         26.9  
...        ...  
3670      25.4  
3671      25.7  
3672      26.1  
3673      26.6  
3674      27.2  

[3532 rows x 8 columns]

### Series
 In pandas, Series is equivalent to vectors in NumPy, representing basic data structures alongside Data Frames, tailored for seamless integration within the broader context of pandas' data-handling functionalities.
* Usage
    * Accessing Columns: To access a column, use the syntax df['column_name']. For instance, c = df['life_expectancy'] stores the 'life_expectancy' column in c.
    * Series as Output: Both rows and columns, when accessed individually, are represented as a Series.

In [ ]:
DF['country']

0       Afghanistan
1       Afghanistan
2       Afghanistan
3       Afghanistan
4       Afghanistan
           ...     
3670       Zimbabwe
3671       Zimbabwe
3672       Zimbabwe
3673       Zimbabwe
3674       Zimbabwe
Name: country, Length: 3675, dtype: object

In [ ]:
pd.Series(DF['country'])

0       Afghanistan
1       Afghanistan
2       Afghanistan
3       Afghanistan
4       Afghanistan
           ...     
3670       Zimbabwe
3671       Zimbabwe
3672       Zimbabwe
3673       Zimbabwe
3674       Zimbabwe
Name: country, Length: 3675, dtype: object

In [ ]:
# Accessing a single column
DF['country']

0       Afghanistan
1       Afghanistan
2       Afghanistan
3       Afghanistan
4       Afghanistan
           ...     
3670       Zimbabwe
3671       Zimbabwe
3672       Zimbabwe
3673       Zimbabwe
3674       Zimbabwe
Name: country, Length: 3675, dtype: object

#### Series Operations



Pandas Series supports various functions like **sum(), min(), max(), count()**, etc., for aggregation.
Performing these functions on 'life_exp' series
 * sum()

In [ ]:
3675.000000*69.849306

256696.19955

In [ ]:
DF.describe()

year     life_exp    hdi_index  co2_consump            gdp  \
count  3675.000000  3675.000000  3563.000000  3671.000000    3633.000000   
mean   2008.000000    69.849306     0.674864     4.712731   11966.053675   
std       6.056125     8.886563     0.164834     6.567435   17105.787953   
min    1998.000000    32.500000     0.255000     0.015900     238.000000   
25%    2003.000000    63.900000     0.537500     0.560500    1470.000000   
50%    2008.000000    71.700000     0.699000     2.250000    4280.000000   
75%    2013.000000    76.400000     0.805000     6.615000   13600.000000   
max    2018.000000    84.800000     0.956000    67.100000  105000.000000   

          services  
count  3675.000000  
mean     51.248705  
std      18.312501  
min       5.590000  
25%      37.600000  
50%      52.900000  
75%      65.700000  
max      88.500000

In [ ]:
 # calculates the total sum of the 'life_expec' Series
 DF['life_exp'].sum()

256696.2

* min()

In [ ]:
DF['life_exp'].min()

32.5

* max()

In [ ]:
DF['life_exp'].max()

84.8

* count()

In [ ]:
DF['life_exp'].count()

3675

* mean()

In [ ]:

DF['life_exp'].mean()

69.84930612244898

In [ ]:
DF['year'].mean()

2008.0

## Understanding loc and iloc

loc
  * Used for accessing rows by labels/indexes.
Works with the row labels, which are the indexes by default in pandas.

* If rows are deleted, loc continues to reference the original labels unless they are explicitly updated.

Let's explore the commands:

In [ ]:
DF

country continent  year  life_exp  hdi_index  co2_consump     gdp  \
0     Afghanistan      Asia  1998      53.3      0.344       0.0522     NaN   
1     Afghanistan      Asia  1999      54.7      0.348       0.0402     NaN   
2     Afghanistan      Asia  2000      54.7      0.350       0.0370     NaN   
3     Afghanistan      Asia  2001      54.8      0.353       0.0376     NaN   
4     Afghanistan      Asia  2002      55.5      0.384       0.0471   333.0   
...           ...       ...   ...       ...        ...          ...     ...   
3670     Zimbabwe    Africa  2014      58.0      0.547       0.8810  1440.0   
3671     Zimbabwe    Africa  2015      58.6      0.553       0.8810  1450.0   
3672     Zimbabwe    Africa  2016      59.2      0.558       0.7710  1430.0   
3673     Zimbabwe    Africa  2017      59.9      0.563       0.8450  1480.0   
3674     Zimbabwe    Africa  2018      60.6      0.569       0.8500  1510.0   

      services  
0         24.4  
1         24.6  
2         24.7  
3         24.7  
4         25.6  
...        ...  
3670      25.4  
3671      25.7  
3672      26.1  
3673      26.6  
3674      27.2  

[3675 rows x 8 columns]

In [ ]:
DF.iloc[2,4]

0.35

In [ ]:
DF.loc[2,"continent"]

'Asia'

In [ ]:
# Accessing a single row
DF.loc[3]

country        Afghanistan
continent             Asia
year                  2001
life_exp              54.8
hdi_index            0.353
co2_consump         0.0376
gdp                    NaN
services              24.7
Name: 3, dtype: object

In [ ]:
# Accessing multiple rows
DF.loc[[1, 10, 100]]

country      continent  year  life_exp  hdi_index  co2_consump  \
1    Afghanistan           Asia  1999      54.7      0.348       0.0402   
10   Afghanistan           Asia  2008      59.2      0.447       0.1520   
100    Argentina  South America  2014      76.5      0.836       4.5600   

         gdp  services  
1        NaN      24.6  
10     415.0      28.2  
100  13600.0      75.3

In [ ]:
# Slicing with loc
DF.loc[1:5]

country continent  year  life_exp  hdi_index  co2_consump    gdp  \
1  Afghanistan      Asia  1999      54.7      0.348       0.0402    NaN   
2  Afghanistan      Asia  2000      54.7      0.350       0.0370    NaN   
3  Afghanistan      Asia  2001      54.8      0.353       0.0376    NaN   
4  Afghanistan      Asia  2002      55.5      0.384       0.0471  333.0   
5  Afghanistan      Asia  2003      56.5      0.393       0.0509  346.0   

   services  
1      24.6  
2      24.7  
3      24.7  
4      25.6  
5      25.9

In [ ]:
DF.loc[3:34,["year","co2_consump"]]

year  co2_consump
3   2001       0.0376
4   2002       0.0471
5   2003       0.0509
6   2004       0.0368
7   2005       0.0515
8   2006       0.0622
9   2007       0.0838
10  2008       0.1520
11  2009       0.2380
12  2010       0.2900
13  2011       0.4060
14  2012       0.3450
15  2013       0.2800
16  2014       0.2530
17  2015       0.2620
18  2016       0.2450
19  2017       0.2470
20  2018       0.2540
21  1998       0.5630
22  1999       0.9560
23  2000       0.9660
24  2001       1.0300
25  2002       1.2000
26  2003       1.3800
27  2004       1.3400
28  2005       1.3800
29  2006       1.2700
30  2007       1.2900
31  2008       1.4600
32  2009       1.4700
33  2010       1.5600
34  2011       1.7900

What is Masking?

In Pandas, masking refers to applying an operation to each individual value in a DataFrame column, resulting in a series of Boolean values.


In [ ]:
DF[DF['year']>2004]

country continent  year  life_exp  hdi_index  co2_consump     gdp  \
7     Afghanistan      Asia  2005      57.6      0.418       0.0515   360.0   
8     Afghanistan      Asia  2006      58.0      0.429       0.0622   368.0   
9     Afghanistan      Asia  2007      58.5      0.447       0.0838   409.0   
10    Afghanistan      Asia  2008      59.2      0.447       0.1520   415.0   
11    Afghanistan      Asia  2009      59.9      0.460       0.2380   492.0   
...           ...       ...   ...       ...        ...          ...     ...   
3670     Zimbabwe    Africa  2014      58.0      0.547       0.8810  1440.0   
3671     Zimbabwe    Africa  2015      58.6      0.553       0.8810  1450.0   
3672     Zimbabwe    Africa  2016      59.2      0.558       0.7710  1430.0   
3673     Zimbabwe    Africa  2017      59.9      0.563       0.8450  1480.0   
3674     Zimbabwe    Africa  2018      60.6      0.569       0.8500  1510.0   

      services  
7         26.5  
8         26.9  
9         27.7  
10        28.2  
11        29.5  
...        ...  
3670      25.4  
3671      25.7  
3672      26.1  
3673      26.6  
3674      27.2  

[2450 rows x 8 columns]

Using .loc with a Mask:

In [ ]:
# DATAFRAME[CONDITION--> Boolean]

# Only the values that are true will be returned..

SyntaxError: ignored

In [ ]:
# masking -> wherever life_exp > 60 will retain
DF[DF.life_exp>60]

country continent  year  life_exp  hdi_index  co2_consump     gdp  \
12    Afghanistan      Asia  2010      60.5      0.472        0.290   547.0   
13    Afghanistan      Asia  2011      61.0      0.477        0.406   533.0   
14    Afghanistan      Asia  2012      61.4      0.489        0.345   580.0   
15    Afghanistan      Asia  2013      61.9      0.496        0.280   592.0   
16    Afghanistan      Asia  2014      61.9      0.500        0.253   588.0   
...           ...       ...   ...       ...        ...          ...     ...   
3650       Zambia    Africa  2015      60.8      0.569        0.271  1340.0   
3651       Zambia    Africa  2016      61.4      0.571        0.296  1350.0   
3652       Zambia    Africa  2017      61.9      0.578        0.304  1350.0   
3653       Zambia    Africa  2018      62.5      0.582        0.302  1370.0   
3674     Zimbabwe    Africa  2018      60.6      0.569        0.850  1510.0   

      services  
12        30.7  
13        31.7  
14        32.8  
15        33.8  
16        34.8  
...        ...  
3650      37.1  
3651      38.0  
3652      38.6  
3653      39.3  
3674      27.2  

[3088 rows x 8 columns]

In [ ]:
# masking -> wherever life_exp > 60 will retain
DF.loc[DF.life_exp>60]

country continent  year  life_exp  hdi_index  co2_consump     gdp  \
12    Afghanistan      Asia  2010      60.5      0.472        0.290   547.0   
13    Afghanistan      Asia  2011      61.0      0.477        0.406   533.0   
14    Afghanistan      Asia  2012      61.4      0.489        0.345   580.0   
15    Afghanistan      Asia  2013      61.9      0.496        0.280   592.0   
16    Afghanistan      Asia  2014      61.9      0.500        0.253   588.0   
...           ...       ...   ...       ...        ...          ...     ...   
3650       Zambia    Africa  2015      60.8      0.569        0.271  1340.0   
3651       Zambia    Africa  2016      61.4      0.571        0.296  1350.0   
3652       Zambia    Africa  2017      61.9      0.578        0.304  1350.0   
3653       Zambia    Africa  2018      62.5      0.582        0.302  1370.0   
3674     Zimbabwe    Africa  2018      60.6      0.569        0.850  1510.0   

      services  
12        30.7  
13        31.7  
14        32.8  
15        33.8  
16        34.8  
...        ...  
3650      37.1  
3651      38.0  
3652      38.6  
3653      39.3  
3674      27.2  

[3088 rows x 8 columns]

In [ ]:
# Accessing rows and columns together
DF.loc[1:5, ['continent', 'country']]

continent      country
1      Asia  Afghanistan
2      Asia  Afghanistan
3      Asia  Afghanistan
4      Asia  Afghanistan
5      Asia  Afghanistan

In [ ]:
DF[DF['co2_consump'] > 0.3]

country continent  year  life_exp  hdi_index  co2_consump     gdp  \
13    Afghanistan      Asia  2011      61.0      0.477        0.406   533.0   
14    Afghanistan      Asia  2012      61.4      0.489        0.345   580.0   
21        Albania    Europe  1998      74.8      0.655        0.563  1600.0   
22        Albania    Europe  1999      75.1      0.665        0.956  1820.0   
23        Albania    Europe  2000      75.4      0.671        0.966  1960.0   
...           ...       ...   ...       ...        ...          ...     ...   
3670     Zimbabwe    Africa  2014      58.0      0.547        0.881  1440.0   
3671     Zimbabwe    Africa  2015      58.6      0.553        0.881  1450.0   
3672     Zimbabwe    Africa  2016      59.2      0.558        0.771  1430.0   
3673     Zimbabwe    Africa  2017      59.9      0.563        0.845  1480.0   
3674     Zimbabwe    Africa  2018      60.6      0.569        0.850  1510.0   

      services  
13        31.7  
14        32.8  
21        33.0  
22        33.8  
23        34.4  
...        ...  
3670      25.4  
3671      25.7  
3672      26.1  
3673      26.6  
3674      27.2  

[3034 rows x 8 columns]

In [ ]:
DF.drop()

ValueError: ignored

In [ ]:
#Dropping a col here
DF.drop('year',axis=1,inplace=True)

In [ ]:
# Deleting a row here
DF.drop(5, inplace=True, axis=0)

In [ ]:
DF.head(10)

country continent  life_exp  hdi_index  co2_consump    gdp  services
0   Afghanistan      Asia      53.3      0.344       0.0522    NaN      24.4
1   Afghanistan      Asia      54.7      0.348       0.0402    NaN      24.6
2   Afghanistan      Asia      54.7      0.350       0.0370    NaN      24.7
3   Afghanistan      Asia      54.8      0.353       0.0376    NaN      24.7
4   Afghanistan      Asia      55.5      0.384       0.0471  333.0      25.6
6   Afghanistan      Asia      57.1      0.409       0.0368  336.0      26.1
7   Afghanistan      Asia      57.6      0.418       0.0515  360.0      26.5
8   Afghanistan      Asia      58.0      0.429       0.0622  368.0      26.9
9   Afghanistan      Asia      58.5      0.447       0.0838  409.0      27.7
10  Afghanistan      Asia      59.2      0.447       0.1520  415.0      28.2

  iloc
 * Used for accessing rows by their integer positions.
Works with the positional indexes, i.e., the numerical position of rows.
 * Reflects the current structure of the DataFrame, adjusting to deletions or changes in the order of rows.

Let's explore the commands:

In [ ]:
# Accessing a single row
DF.iloc[3]

country        Afghanistan
continent             Asia
life_exp              54.8
hdi_index            0.353
co2_consump         0.0376
gdp                    NaN
services              24.7
Name: 3, dtype: object

In [ ]:
# Accessing multiple rows
DF.iloc[[1, 10, 100]]

country      continent  life_exp  hdi_index  co2_consump      gdp  \
1    Afghanistan           Asia      54.7      0.348       0.0402      NaN   
11   Afghanistan           Asia      59.9      0.460       0.2380    492.0   
101    Argentina  South America      76.5      0.840       4.6400  13800.0   

     services  
1        24.6  
11       29.5  
101      76.1

In [ ]:
# Slicing with iloc
DF.iloc[1:5]

country continent  life_exp  hdi_index  co2_consump    gdp  services
1  Afghanistan      Asia      54.7      0.348       0.0402    NaN      24.6
2  Afghanistan      Asia      54.7      0.350       0.0370    NaN      24.7
3  Afghanistan      Asia      54.8      0.353       0.0376    NaN      24.7
4  Afghanistan      Asia      55.5      0.384       0.0471  333.0      25.6

In [ ]:
# Accessing rows and columns together
DF.iloc[1:5, 0:3]

country continent  life_exp
1  Afghanistan      Asia      54.7
2  Afghanistan      Asia      54.7
3  Afghanistan      Asia      54.8
4  Afghanistan      Asia      55.5

## Basic Operations in Pandas



1. Accessing Columns:
 * To access a column in a DataFrame, use the syntax df['column_name'].
 * This returns the specified column as a Series.

In [ ]:
#extracts the 'life_expectancy' column from DF and stores it in c.
c = DF['life_exp']
# returned series is stored and printed via c variable
c

0       53.3
1       54.7
2       54.7
3       54.8
4       55.5
        ... 
3670    58.0
3671    58.6
3672    59.2
3673    59.9
3674    60.6
Name: life_exp, Length: 3674, dtype: float64

2. Calculating Mean of a Column:
* Use the mean() method on a Series to calculate its mean value.
Example: DF.mean() computes the mean of the 'life_exp' Series.


In [ ]:
DF.mean()

<ipython-input-85-86b3940b9e7d>:1: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  DF.mean()


life_exp          69.852940
hdi_index          0.674943
co2_consump        4.714002
gdp            11969.253029
services          51.255604
dtype: float64

* To find the mean of a specific column, you can reference it by name:


In [ ]:
DF['life_exp'].mean()

69.85293957539467

### Filtering Data


* Conditional Filtering:
 * Apply conditions directly to DataFrame columns for filtering.





In [ ]:
DF[(DF['life_exp']>30) & (df['life_exp']<40)]

<ipython-input-87-5cb0438efaba>:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  DF[(DF['life_exp']>30) & (df['life_exp']<40)]


country      continent  life_exp  hdi_index  co2_consump     gdp  \
1419   Haiti  North America      32.5      0.471        0.211  1290.0   

      services  
1419      59.9

In [ ]:
df[(df['life_exp']>30)&(df['life_exp']<40)]

country      continent  year  life_exp  hdi_index  co2_consump     gdp  \
1419   Haiti  North America  2010      32.5      0.471        0.211  1290.0   

      services  
1419      59.9

In [ ]:
#filters the DataFrame to only include rows where life expectancy is greater than 40.
DF[DF['life_exp'] > 40]

country continent  life_exp  hdi_index  co2_consump     gdp  \
0     Afghanistan      Asia      53.3      0.344       0.0522     NaN   
1     Afghanistan      Asia      54.7      0.348       0.0402     NaN   
2     Afghanistan      Asia      54.7      0.350       0.0370     NaN   
3     Afghanistan      Asia      54.8      0.353       0.0376     NaN   
4     Afghanistan      Asia      55.5      0.384       0.0471   333.0   
...           ...       ...       ...        ...          ...     ...   
3670     Zimbabwe    Africa      58.0      0.547       0.8810  1440.0   
3671     Zimbabwe    Africa      58.6      0.553       0.8810  1450.0   
3672     Zimbabwe    Africa      59.2      0.558       0.7710  1430.0   
3673     Zimbabwe    Africa      59.9      0.563       0.8450  1480.0   
3674     Zimbabwe    Africa      60.6      0.569       0.8500  1510.0   

      services  
0         24.4  
1         24.6  
2         24.7  
3         24.7  
4         25.6  
...        ...  
3670      25.4  
3671      25.7  
3672      26.1  
3673      26.6  
3674      27.2  

[3673 rows x 7 columns]

* For multiple conditions, combine them using & (AND) and | (OR), with each condition enclosed in parentheses.

In [ ]:
#filters for life expectancy between 30 and 40.
df[(df['life_exp'] > 30) & (df['life_exp'] <= 40)]

country      continent  year  life_exp  hdi_index  co2_consump     gdp  \
1419   Haiti  North America  2010      32.5      0.471        0.211  1290.0   

      services  
1419      59.9

###Grouping Data


In [ ]:
DF.groupby('country')['life_exp'].min()

country
Afghanistan    53.3
Albania        74.8
Algeria        70.2
Angola         50.6
Argentina      73.7
               ... 
Vanuatu        63.0
Vietnam        71.7
Yemen          62.5
Zambia         45.2
Zimbabwe       46.4
Name: life_exp, Length: 175, dtype: float64

* GroupBy Functionality:
  * The groupby() method groups DataFrame rows based on a column's values.
  * After grouping, you can apply aggregation functions like mean(), sum(), etc.



In [ ]:
#Example:  calculates the average life expectancy for each year.
DF.groupby('year')['life_exp'].mean()

KeyError: ignored

  * Counting Unique Values:
nunique() method counts the number of unique values in a column.


In [ ]:
#counts the number of unique countries in each continent.
DF.groupby('continent')['country'].nunique()

In [ ]:
#counts the number of unique countries in each continent.
DF.groupby('year')[['life_exp','gdp']].count()

The above output represents the count of unique countries within each continent group.


* Multi-Level Grouping:
 * Group by multiple columns by passing a list of column names to groupby().
 * This creates hierarchical indices based on the specified columns.



In [ ]:
DF.groupby(['continent','year'])[['life_exp','gdp']].mean()

In [ ]:
#groups the data first by 'continent' and then by 'year', calculating the mean for each group.
DF.groupby(['continent', 'year']).mean()

In [ ]:
13 DECEMBER'2023'

In [ ]:
DF

In [ ]:
#Other syntax to aggregate in groupby
DF.groupby(['']).agg({})

In [ ]:
DF

In [ ]:
DF.groupby('continent').max()['']

In [ ]:
DF.groupby(["continent","year"]).agg({'life_exp':'mean','co2_consump':'max'})

In [ ]:
merged_df.agg({'status_x':['sum','count']})

### Sorting Data
* Basic Sorting: df.sort_values('life_exp') sorts the DataFrame based on the 'life expectancy' column in ascending order.



In [ ]:
df.sort_values('co2_consump',ascending=False).sort_values("country",ascending=True)

In [ ]:
df.sort_values(["co2_consump","gdp"],ascending=[False,True])

In [ ]:
df.sort_values('life_exp')

* Changing Sort Order: Adding **ascending=False** sorts in descending order.


In [ ]:
df.sort_values('life_exp',ascending=False)

* Multi-Level Sorting: Sorting can be done on multiple columns.

In [ ]:
DF.sort_values(by=['country', 'year'])

In [ ]:
#sorts primarily by 'life expectancy' and then by 'year' within each 'life expectancy' group.

DF.sort_values(by=['country', 'year'],ascending=[True,False])

* Sorting Behavior: If the primary sort column (like 'life expectancy') has unique values for each row, the secondary sort column (like 'year') may not influence the order.

In [ ]:
# Sort the DataFrame first by 'Life_Exp' in ascending order,
# and then by 'Year' in ascending order for ties in 'Life_Exp'

DF.sort_values(by=['life_exp', 'year'])


* Sorting with Different Orders: You can specify different sort orders for each column by passing a list to ascending.

In [ ]:
#sorts the first column in descending and the second in ascending order.
DF.sort_values(by=['life_exp', 'year'], ascending=[True, False])

### String Operations

In [ ]:
DF['country']

In [ ]:
DF['country']>'AAA'

* String Comparison:
 * Pandas supports string comparison operations, useful for filtering.



In [ ]:
DF[DF['co2_consump']>0.3]

In [ ]:
#filters the DataFrame for countries alphabetically before 'Nigeria'.
DF[DF['country'] < 'Nigeria']

### Resetting Index

  * After operations like filtering or grouping, the DataFrame's index may need to be reset for consistency.
  * Use DF.reset_index() to reset the index of the DataFrame.
  * This method is also useful when indices become non-sequential after row deletions.

In [ ]:
xyz=DF.drop(2,axis=0)

In [ ]:
xyz.reset_index(drop=True)

In [ ]:
DF.set_index('country').reset_index()

In [ ]:
#Reset the index
DF.reset_index()

##Restructuring Data with Melt
Using **pd.melt**:

 * Method Application: Showed how to use pd.melt to restructure the DataFrame. This process involved merging multiple 'day' columns into a single 'day' column and creating a corresponding 'temp' column for temperatures.

 * Resulting Format: The outcome was a more streamlined DataFrame with consolidated day and temperature data.

Understanding **pd.melt** Parameters:
   * **id_vars:** Explained the use of id_vars to specify columns that should remain unchanged during the melt process.
   * **var_name and value_name**: Clarified the roles of var_name (to store the name of the melted columns) and value_name (to store the corresponding values).


In [ ]:
import pandas as pd
# Use the correct raw URL for the CSV file on GitHub
csv_url = 'https://raw.githubusercontent.com/yesdhruv/csv_files/master/weather.csv'


weather = pd.read_csv(csv_url)
# Display the DataFrame
weather

In [ ]:
# Perform the melt operation
weather_melt = pd.melt(weather,
                       id_vars=['MinTemp', 'MaxTemp', 'Rainfall'],  # Include actual columns in your DataFrame
                       var_name='day',
                       value_name='temp')

In [ ]:
weather_melt[weather_melt['MinTemp']==8.0]

In [ ]:
weather_melt

In [ ]:

print(weather_melt.shape)
# Display the first few rows of the melted DataFrame
weather_melt.head()


In [ ]:
weather_melt.columns

In [ ]:
weather_melt.groupby(['day']).agg({'Rainfall':'first'})

In [ ]:
weather_melt

In [ ]:
# Pivot the melted DataFrame back to its original format
weather_pivot = weather_melt.pivot_table(index=['Rainfall', 'MinTemp', 'MaxTemp'],
                                         columns='day',
                                         values='temp',
                                         aggfunc='first')



# Display the pivoted DataFrame
weather_pivot.head()

In [ ]:
weather_pivot.reset_index()

In [ ]:
# Reset the index to make it look like the original DataFrame
weather_pivot.reset_index(inplace=True)

weather_pivot.head()

In [ ]:
weather_melt

In [ ]:

weather_melt.pivot_table(index=['Rainfall', 'MinTemp', 'MaxTemp'],
                                         columns='day',
                                         values='temp',
                                         aggfunc='first',
                                         dropna=False).reset_index()



In [ ]:

weather_melt.pivot_table(index=['Rainfall', 'MinTemp', 'MaxTemp'],
                                         columns='day',
                                         values='temp',
                                         aggfunc='first',
                                         dropna=True).reset_index()


##Applying Custom Functions in Pandas

In [ ]:
df[['hdi_index','life_exp']].apply(lambda x:x+2)

In [ ]:
df

**.apply()**: This method is used to apply a function along the axis of a DataFrame or Series.

* Axis Specification in .apply():
 * When Axis=0: The function collapses the rows and applies mean_func to each column separately. This results in the mean of each column being calculated.

 * When Axis=1: The function collapses the columns and applies mean_func to each row. This results in the mean of the values in each row being calculated.


A custom function named **mean_func** is created to demonstrate the application of functions in Pandas. This function was designed to calculate the mean of a given set of values.

The function takes an input x and returns the mean calculated as the sum of x divided by the length of x.


In [ ]:
pd.Series([1,2,3,4,5])

In [ ]:
pd.DataFrame([[1,2,3,4],[5,6,7,8]],columns=['A','B','C','D'])

In [ ]:
pd.DataFrame({'A':[1,2,3,4,5],'B':[6,5,4,3,2]})

In [ ]:
import pandas as pd

def mean_func(x):
    """
    Calculate the mean of a given set of values.

    Parameters:
    - x: Input values for which the mean is calculated.

    Returns:
    - Mean value calculated as the sum of x divided by the length of x.
    """
    # Convert the values to numeric to handle mixed data types
    x_numeric = pd.to_numeric(x, errors='coerce')
    return x_numeric.sum() / len(x_numeric)

# Creating a sample DataFrame
data = {'Day': ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday'],
        'Min_Temperature': [18, 20, 22, 19, 17],
        'Max_Temperature': [25, 28, 24, 26, 23]}
df = pd.DataFrame(data)
df

In [ ]:
43/3

In [ ]:
df.apply(mean_func,axis=1)

In [ ]:
# Applying mean_func with axis=0 (Column-wise mean)
column_means = df.apply(mean_func, axis=0)


# Displaying the results
print("Column-wise Mean:")
print(column_means)

In [ ]:
# Applying mean_func with axis=1 (Row-wise mean)
row_means = df.apply(mean_func, axis=1)

print("\nRow-wise Mean:")
print(row_means)

# Creating Series and DataFrames from Scratch


## Creating a Series


* Series Constructor: pd.Series() is used to create a  Series. Remember, Series starts with an uppercase 'S', indicating it's a constructor.



In [ ]:
# creates a Series with values 10, 20, and 30.
pd.Series([10, 20, 30,[40,50,40]])

## Creating a DataFrame
DataFrame Constructor: pd.DataFrame() is used for creating a DataFrame.




* Row-Based Creation: Passing a list of lists, where each inner list represents a row.

In [ ]:
#creates a Pandas DataFrame with two columns, 'A' and 'B', and two rows.
pd.DataFrame([[10,20],[30,40]],columns=['A','B'])



* Column-Based Creation: Using a dictionary where keys are column names and values are lists of column data.



In [ ]:

pd.DataFrame({'A': [10, 30], 'B': [20, 40]})

* Adding a Single Row: To add a single row, it must be wrapped in an additional list to represent a list of rows.

 Lets create two sample data frames for further operations

In [ ]:
# Creating df1
df1 = pd.DataFrame({'A':[10, 30], 'B':[20, 40]})

df1

In [ ]:
# Creating df2
df2 = pd.DataFrame({'A':[10, 30], 'C':[20, 40]})

df2

In [ ]:
df1.append(df2)

In [ ]:
pd.concat([df1,df2],axis=0)

#df1.append(df2)

In [ ]:
pd.concat([df1,df2],axis=1)

## Concatenating a DataFrame
* Vertical Concatenation
 * Default Concatenation: Using pd.concat([df1, df2]) concatenates DataFrames vertically. Columns with the same name are aligned together.


In [ ]:
pd.concat([df1, df2])

* Horizontal Concatenation
 * Side-by-Side Concatenation: Setting axis=1 in pd.concat() concatenates DataFrames side by side.


In [ ]:
 pd.concat([df1, df2], axis=1)

#Dealing with Missing Values
Let's load the dataset first, we can upload the data or can use the raw public link.


In [ ]:
df=df.drop(2,axis=0)

or

df.drop(2,axis=0,inplace=True)

In [ ]:
df

In [ ]:
import pandas as pd

In [ ]:
csv_url = 'https://raw.githubusercontent.com/yesdhruv/csv_files/master/heart.csv'
df = pd.read_csv(csv_url)
# Display the DataFrame
df

In [ ]:
df.isnull().sum()

In [ ]:
df.info()

In [ ]:
# pd.read_csv()

### General knowledge on Reading CSV Files
 * CSV Format: Stands for Comma-Separated Values. A common format for storing tabular data where each value is separated by a comma.
 * Reading CSV in Pandas: Use pd.read_csv('url/file_path') to read a CSV file into a DataFrame.
 * Delimiter in CSV: The default delimiter is a comma. For tab-separated values (TSV), use delimiter='\t'.


### Understanding Missing Values
 * Missing Values in DataFrames
   * Represented as NaN, indicating data is not available or not applicable.
 * Importance
   * Handling missing values is crucial before feeding data into machine learning algorithms.


* Applying isnull() and notnull() to the Entire DataFrame:
This will give you a DataFrame where each cell is either True (if missing) or False.


In [ ]:
# Using isnull() to identify missing values
missing_values = df.isnull()

# Using notnull() to identify non-missing values
non_missing_values = df.notnull()

In [ ]:
non_missing_values

In [ ]:
# Assuming your DataFrame is named 'df'
# Count the number of zero values in each column
zero_counts = (df == 0).sum()

In [ ]:


# Display the count of zero values for each column
print(zero_counts)


In [ ]:
# Count the total number of zero values in the entire DataFrame
total_zero_count = (df == 0).sum().sum()

# Display the total count of zero values
print(f'Total number of zero values: {total_zero_count}')


In [ ]:
u= df['age'].value_counts()
len(u)

In [ ]:
df['age'].unique()

## Strategies for Handling Missing Data

Two primary strategies for handling missing data are deletion and imputation.





### Deletion Methods

* Listwise Deletion
 * Involves removing entire rows from a dataset if any value in the row is missing. This is a straightforward approach but can lead to significant data loss.

* Pairwise Deletion
 * Only the specific missing data point is omitted, not the entire row. Useful in statistical analyses like correlation, utilizing the remaining available data.




In [ ]:
# Import NumPy and Pandas
import numpy as np
import pandas as pd

# Create a sample DataFrame with missing values
data = {
    'Name': ['Alice', 'Bob', 'Ali', 'David', 'Eva'],
    'Age': [25, 30, np.nan, 22, 28],
    'Salary': [50000, np.nan, 60000, 400,np.nan],
    'Department': ['HR', 'IT', 'Finance', np.nan, 'Marketing']
}

# Convert the dictionary to a DataFrame
df = pd.DataFrame(data)
df


In [ ]:
# Deletion: Listwise Deletion
df_listwise = df.dropna()
print("\nDataFrame after Listwise Deletion:")
df_listwise
# print(df_listwise)

In [ ]:
# Deletion: Pairwise Deletion
df_pairwise = df.dropna(subset=['Name', 'Salary'])
print("\nDataFrame after Pairwise Deletion:")
df_pairwise

In [ ]:
df.fillna(9)

 Pros and Cons of Deletion
- Straightforward and easy to implement.
- Can introduce bias, affecting the representativeness and reliability of the dataset.
- Best suited for scenarios with minimal missing data or when the missing data is randomly distributed.

### Imputation Methods

* Mean/Median/Mode Imputation
 * Involves replacing missing values with the mean, median, or mode of the respective column. Easy to implement and useful for quickly addressing missing data.

<!-- * K-Nearest Neighbors (KNN) Imputation
 * Utilizes the nearest neighbors of a missing data point to predict and impute its value. More sophisticated and can be more accurate than mean/median/mode imputation.

* Multiple Imputation
 * Involves creating multiple sets of imputations for missing values, providing a comprehensive approach by addressing the uncertainty inherent in any single imputation. -->

#### Pros and Cons of Imputation
- Helps in preserving the original size of the dataset.
- Can introduce inaccuracies or biases if not implemented with an understanding of the dataset's characteristics.
- Requires careful consideration of the nature of the missing data and the overall dataset structure.
- Often preferred over deletion methods in scenarios where retaining the maximum amount of data is crucial.

In [ ]:
# Imputation: Mean Imputation
df_mean_imputed = df.fillna(df.mean())
print("\nDataFrame after Mean Imputation:")
print(df_mean_imputed)

In [ ]:
import pandas as pd
import numpy as np
from sklearn.impute import KNNImputer

# Sample DataFrame with missing values
data = {'A': [1, 2, np.nan, 4, 5],
        'B': [10, 20, 30, np.nan, 50],
        'C': [100, 200, 300, 400, 500]}

df = pd.DataFrame(data)


In [ ]:
# Check for missing values in the DataFrame
if df.isnull().values.any():
    # Handling missing values by filling with the mean
    df_filled = df.fillna(df.mean())

    # Check if there are still any missing values
    if df_filled.isnull().values.any():
        raise ValueError("DataFrame still contains missing values. Check and handle missing values appropriately.")
else:
    df_filled = df.copy()  # No missing values, create a copy of the original DataFrame


In [ ]:
# Check if the DataFrame contains numeric data
numeric_columns = df_filled.select_dtypes(include=[np.number]).columns
if len(numeric_columns) == 0:
    raise ValueError("DataFrame does not contain numeric columns. KNN imputation requires numeric data.")

# Imputation: K-Nearest Neighbors (KNN) Imputation
imputer = KNNImputer(n_neighbors=2)
df_knn_imputed = pd.DataFrame(imputer.fit_transform(df_filled), columns=df_filled.columns)


In [ ]:
# Displaying Results
print("\nOriginal DataFrame:")
print(df)

print("\nDataFrame after Handling Missing Values:")
print(df_filled)

print("\nDataFrame after KNN Imputation:")
print(df_knn_imputed)


In [ ]:
df

#Exploratory Data Analysis (EDA)
 - EDA involves examining and understanding the dataset's characteristics.
 - Common EDA tasks include cleaning data, handling missing values, and adding useful columns.
 - Example: Converting timestamps to datetime objects in Pandas for easier manipulation.


Importing the dataset of uber


In [ ]:
import pandas as pd

url = 'https://raw.githubusercontent.com/yesdhruv/csv_files/master/Uber%20Request%20Data.csv'
data = pd.read_csv(url)
data


In [ ]:
type(data)

In [ ]:
data.shape

In [ ]:
data.columns

###Data Cleaning and Manipulation

In [ ]:
data

* Dropping Specific Rows
 * data.drop Method: Used for removing specific rows from a DataFrame in pandas.

* Labels vs. Index:
 * Labels: Drop rows using index labels (non-numerical).
 * Index: Drop rows by specifying their numerical index.

In [ ]:
print(data.index)


In [ ]:
# Assuming 'Request id' is the current index and you want to drop rows with specific labels
labels_to_drop = [0,1,2,4,6,7,8,9,10,15,6735, 6736, 6737, 6738, 6739, 6740, 6741, 6742, 6743, 6744]

# Drop rows based on labels using the 'labels' parameter
data.drop(labels=labels_to_drop, axis=0, inplace=False)

data

In [ ]:
print(data.index)

In [ ]:
data.drop(index=[3, 5], axis=0, inplace=True)
data

In [ ]:
import pandas as pd
url = 'https://raw.githubusercontent.com/yesdhruv/csv_files/master/Uber%20Request%20Data.csv'

data = pd.read_csv(url)
data

* Renaming Column Names
 * Removing Unwanted Characters





In [ ]:
# Mapping:
# Original column names
original_columns = data.columns

# Define a mapping of old column names to new column names
column_mapping = {
    'Request id': 'reqUEST_Id',
    'Pickup point': 'pickUP_POINT',
    'Driver id': 'driVER_Id',
    'Status': 'sta-TUS',
    'Request timestamp': 'reqUEST_timeSTAMP',
    'Drop timestamp': 'drOP_timeSTAMP'
}

# Rename columns using the mapping
data.rename(columns=column_mapping, inplace=True)

updated_columns = data.columns

print("Updated Columns:",updated_columns)


In [ ]:
# List Comprehension to remove asterisks from column names
data.columns = [col.replace('_', '+^^+') for col in data.columns]

# Display the DataFrame with updated column names
data


In [ ]:
#Custom Function:
def remove_special_chars(column_name):
    return column_name.replace('reqUEST+^^+Id', 'reqID')

data.columns = [remove_special_chars(col) for col in data.columns]
# Display the DataFrame with updated column names
data

###Working with DateTime

* pd.to_datetime Function: Converts a column containing date values into DateTime objects.


DateTime objects provide convenient methods to extract various attributes like year, month, day, day name, etc.

In [ ]:
import pandas as pd
url = 'https://raw.githubusercontent.com/yesdhruv/csv_files/master/Uber%20Request%20Data.csv'

data = pd.read_csv(url)
data

In [ ]:
pd.to_datetime(data['Request timestamp'])

In [ ]:
data.info()

In [ ]:
data['Request timestamp']= pd.to_datetime(data['Request timestamp'])

In [ ]:
data.info()